### PR TITLE
Progressive delivery of applications using ApplicationSets

### PR DESCRIPTION
**Pull request by iam-veeramalla**
**What type of PR is this?**
> /kind enhancement

**What does this PR do / why we need it**:
This is an experimental, alpha-quality feature that allows you to control the order in which the ApplicationSet controller will create or update the Applications owned by an ApplicationSet resource. It may be removed in future releases or modified in backwards-incompatible ways.

The feature only interacts with the health of managed Applications. It is **NOT** intended to support direct integrations with other Rollout controllers (such as the native ReplicaSet controller or Argo Rollouts).
Progressive Rollouts watch for the managed Application resources to become "Healthy" before proceeding to the next stage.

**Enabling Progressive Rollouts**
As an experimental feature, progressive rollouts must be explicitly enabled, in one of the below 3 ways.
1. Pass `--enable-progressive-rollouts` to the ApplicationSet controller args.
    To achieve this using the operator, Users can add the above argument in the Argo CD CR
```
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: example-argocd
  labels:
    example: basic
spec:
  applicationSet:
    extraCommandArgs:
     - --enable-progressive-rollouts
```
2. Set `ARGOCD_APPLICATIONSET_CONTROLLER_ENABLE_PROGRESSIVE_ROLLOUTS=true` in the ApplicationSet controller environment variables in the below mentioned way.
```
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: example-argocd
  labels:
    example: basic
spec:
  applicationSet:
    env:
     - name: ARGOCD_APPLICATIONSET_CONTROLLER_ENABLE_PROGRESSIVE_ROLLOUTS
       value: "true"
```
3. Set `applicationsetcontroller.enable.progressive.rollouts: true` in the ArgoCD `ConfigMap`.
```
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: example-argocd
  labels:
    example: basic
spec:
   extraConfig:
     applicationsetcontroller.enable.progressive.rollouts: true
```

**Have you updated the necessary documentation?**
* [x] Documentation update is required by this PR.
* [x] Documentation has been updated.

**How to test changes / Special notes to the reviewer**:
1. Run the operator manually using `make install run`
2. Once the operator is running, submit any one of the 3 Argo CD CRs mentioned above.
3. The expected behavior for each of the example is provided above.

### CODE REVIEW

**Comment by iam-veeramalla** [comment]
> ting that fiel

Hi @jaideepr97 , the implementation is same as upstream.

Upstream also supports all the three ways and three are completely unique

**Comment by iam-veeramalla** [comment]
@jaideepr97 fyi
https://argo-cd.readthedocs.io/en/latest/operator-manual/applicationset/Progressive-Rollouts/#enabling-progressive-rollouts

**Comment by jaideepr97** [comment]
@iam-veeramalla I understand that 
I'm saying as an operator I think we can try and make things more user friendly 
what if we just had a toggle under `.spec.applicationSets.enableProgressiveRollout` that set all these fields to true under the hood?

**Comment by iam-veeramalla** [comment]
> @iam-veeramalla I understand that I'm saying as an operator I think we can try and make things more user friendly what if we just had a toggle under `.spec.applicationSets.enableProgressiveRollout` that set all these fields to true under the hood?

@jaideepr97 Thanks for the suggestion.
The idea here is to implement the `extraCommandArgs` and `env` support so that in the future, we don't need to add more fields to the Argo CD CR w.r.t command arguments or env variable for ApplicationSets.
It might look like we are adding too many place holders for this PR but we are actually not. The applicationset controller lacks these options which are already supported by `controller`, `server`, `repo-server` and other components.

File: examples/argocd-route.yaml [hunk 1]
```diff
@@ -5,6 +5,7 @@ metadata:
   labels:
     example: route
 spec:
+  applicationSet: {}
```
Comment (jaideepr97) [inline examples/argocd-route.yaml:8]: @iam-veeramalla is this intentionally added here to the example route?

Comment (iam-veeramalla) [reply examples/argocd-route.yaml:8]: ah no, good catch. I will get rid of it.

**Comment by jaideepr97** [comment]
@iam-veeramalla could we add unit tests as well?

**Comment by iam-veeramalla** [comment]
@jaideepr97 added new tests and removed an unintentional change. PTAL, Thanks :)

File: docs/reference/argocd.md [hunk 2]
```diff
@@ -97,6 +99,23 @@ spec:
...
+metadata:
+  name: example-argocd
+  labels:
+    example: applicationset
+spec:
+  applicationSet:
+    extraCommandArgs:
+      - --foo
+      - bar
+```
```
[the remark this answers is not in the document]
Comment (iam-veeramalla) [reply docs/reference/argocd.md:118]: I think we should keep it very generic. In the release notes text when we describe the feature, we can add a point for this.

**Review by jaideepr97 — APPROVED** [review]

**Pull request merged**